specification/dartLangSpec.tex

```diff
@@ -2521,49 +2521,76 @@ \section{Classes}
-and $C$ does not declare a member with the same name as $m$.
+Assume that $I$ has an accessible member signature $m$.
 It is a compile-time error if $C$ does not have
 a concrete accessible member with the same name as $m$,
 unless $C$ has a non-trivial \code{noSuchMethod}
 (\ref{theMethodNoSuchMethod}).
-It is a compile-time error if $C$ has
-a concrete accessible member with the same name as $m$,
-with a method signature $m'$ which is not a correct override of $m$
+
+\LMHash{}%
+Each concrete member must have a suitable signature.
+Assume that $C$ has a concrete accessible member with the same name as $m$,
+and let $m''$ be its member signature.
+\commentary{%
+The concrete member may be declared in $C$ or inherited from a superclass.%
+}
+Let \DefineSymbol{m'} be the member signature which is obtained from $m''$
+by adding, if not present already, the modifier \COVARIANT{}
+(\ref{covariantParameters})
+to each parameter $p$ in $m''$ where
+the corresponding parameter in $m$ has the modifier \COVARIANT.
+It is a compile-time error if $m'$ is not a correct override of $m$
 (\ref{correctMemberOverrides}),
 unless that concrete member is a \code{noSuchMethod} forwarder
 (\ref{theMethodNoSuchMethod}).
 
 \commentary{%
-In particular, it is an error for a class to be concrete even if it inherits
+So it is an error for a class to be concrete even if it declares or inherits
 a member implementation for every member signature in its interface,
 unless each of them has parameters and types such that they satisfy
 the corresponding member signature.
-But when there is a non-trivial \code{noSuchMethod} it is allowed
+For this check, any missing \COVARIANT{} modifiers are implicitly added
+to the inherited signature.
+When the modifier \COVARIANT{} is added to one or more parameters
+(which will only happen when the concrete member is inherited),
+an implementation may choose to implicitly induce a forwarding method
+with the same signature as $m'$,
+in order to perform the required dynamic type check
+and then invoke the inherited method.
+
+When there is a non-trivial \code{noSuchMethod} it is allowed
 to leave some members unimplemented,
 and it is allowed to have a \code{noSuchMethod} forwarder which does not
 satisfy the class interface
 (in which case it will be overridden by another \code{noSuchMethod} forwarder).%
 }
 
 \LMHash{}%
-Assume that $C$ has a concrete accessible member with the same name as $m$,
-with a method signature $m'$ which is a correct override of $m$.
-For each parameter $p$ of $m'$ where \COVARIANT{} is present,
-it is a compile-time error if there exists
-a direct or indirect superinterface of $C$ which has
-an accessible method signature $m''$ with the same name as $m$,
-such that $m''$ has a parameter $p''$ that corresponds to $p$
-(\ref{covariantParameters}),
-unless the type of $p$ is a subtype or a supertype of the type of $p''$.
+One more constraint must be satisfied
```

Comment:
Nice foreshadowing, but be more direct. :)

> Parameters that are covariant-by-declaration must also satisfy the following constraint:

Reply:
Done.